Limber/MIXMockServer.cpp

```diff
@@ -0,0 +1,263 @@
+#include <Swiften/MIX/MIX.h>
+#include <Swiften/Network/BoostConnection.h>
+#include <Swiften/Network/BoostConnectionServer.h>
+#include <Swiften/Network/BoostIOServiceThread.h>
+#include <Swiften/Network/ConnectionServer.h>
+#include <Swiften/Roster/XMPPRosterImpl.h>
+#include <Swiften/Parser/PayloadParsers/FullPayloadParserFactoryCollection.h>
+#include <Swiften/Parser/PlatformXMLParserFactory.h>
+#include <Swiften/Serializer/PayloadSerializers/FullPayloadSerializerCollection.h>
+
+#include <Limber/Server/ServerFromClientSession.h>
+#include <Limber/Server/SimpleUserRegistry.h>
+
+#include <Swiften/Base/Log.h>
+
+using namespace Swift;
+
+class SimpleMIXChannelRegistry {
+
+    public:
+        SimpleMIXChannelRegistry() {}
+
+        void addMIXChannel(const JID& channelJID) {
+            channels_.insert(channelJID);
+        }
+
+        const std::unordered_set<std::string>& getChannels() {
+            return channels_;
+        }
+
+        bool hasMIXChannel(const JID& jid) {
+            return channels_.find(jid) != channels_.end();
+        }
+
+    private:
+        std::unordered_set<std::string> channels_;
+};
+
+class Server {
+    public:
+        Server(UserRegistry* userRegistry, SimpleMIXChannelRegistry* mixChannelRegistry, EventLoop* eventLoop) : userRegistry_(userRegistry), mixChannelRegistry_(mixChannelRegistry) {
+            serverFromClientConnectionServer_ = BoostConnectionServer::create(5222, boostIOServiceThread_.getIOService(), eventLoop);
+            serverFromClientConnectionServer_->onNewConnection.connect(boost::bind(&Server::handleNewConnection, this, _1));
+            serverFromClientConnectionServer_->start();
+        }
+
+    private:
+        void handleNewConnection(std::shared_ptr<Connection> c) {
+            auto session  = std::make_shared<ServerFromClientSession>(idGenerator_.generateID(), c, &payloadParserFactories_, &payloadSerializers_, &xmlParserFactory, userRegistry_);
+            serverFromClientSessions_.push_back(session);
+            session->onElementReceived.connect(boost::bind(&Server::handleElementReceived, this, _1, session));
+            session->onSessionFinished.connect(boost::bind(&Server::handleSessionFinished, this, session));
+            session->startSession();
+        }
+
+        void handleSessionFinished(std::shared_ptr<ServerFromClientSession> session) {
+            serverFromClientSessions_.erase(std::remove(serverFromClientSessions_.begin(), serverFromClientSessions_.end(), session), serverFromClientSessions_.end());
+        }
+
+        void handleElementReceived(std::shared_ptr<ToplevelElement> element, std::shared_ptr<ServerFromClientSession> session) {
+            std::shared_ptr<Stanza> stanza(std::dynamic_pointer_cast<Stanza>(element));
+            if (!stanza) {
+                return;
+            }
+            stanza->setFrom(session->getRemoteJID());
+
+            auto iq = std::dynamic_pointer_cast<IQ>(stanza);
+            if (iq) {
```

Comment:
It probably makes more sense to test for `!iq` and handle the error case there and afterwards check if it's a roster payload.